File: articles_sdk_ts/resources.ts
import { Collection } from "./_runtime/collection.js";
import { Resource } from "./_runtime/resource.js";

export interface ArticleQuery {
    title__contains?: string;
    category?: number | null;
}

export interface ArticleGetQuery {}

export interface ArticleEdit {
    title?: string | null;
    content?: string | null;
    author?: User | number | null;
}

export interface ArticleCreate {
    title?: string;
    content?: string;
    author?: User | number | null;
    categories?: (Category | number[])[] | null;
}

export interface UserQuery {
    username?: string;
}

export interface UserGetQuery {}

export interface CategoryQuery {
    article?: number | null;
}

export interface CategoryCreate {
    name?: string;
}

export class ArticleCollection extends Collection<Article> {
  filter(kwargs: Partial<ArticleQuery>): this {
    return super.filter(kwargs as Record<string, unknown>);
  }
  sort(...fields: ('title' | '-title' | 'created_at' | '-created_at')[]): this {
    return super.sort(...fields);
  }
  fields(params: { articles?: ('id' | 'title' | 'content' | 'created_at' | 'author' | 'categories')[] | null }): this {
    const converted: Record<string, string[]> = {};
    for (const [k, v] of Object.entries(params)) {
      if (v != null) converted[k] = v;
    }
    return super.fields(converted);
  }
}
export class Article extends Resource {
      static _type = "articles";
      static _attributeTypes: Record<string, string> = {
        'id': 'number',
        'title': 'string',
        'content': 'string',
        'created_at': 'string',
      };
      static _relationshipTypes: Record<string, [string, boolean]> = {
        'author': ['users', false],
        'categories': ['categories', true],
      };
      static _capabilities = new Set(['create', 'delete', 'edit', 'get_many', 'get_one'] as const);
      get title(): string { return this.get('title') as string; }
      set title(value: string) { this.set('title', value); }
      get content(): string { return this.get('content') as string; }
      set content(value: string) { this.set('content', value); }
      get created_at(): string { return this.get('created_at') as string; }
      set created_at(value: string) { this.set('created_at', value); }
      get author(): User { return this.get('author') as User; }
      set author(value: User) { this.set('author', value); }
      get categories(): CategoryCollection | null { return this.get('categories') as CategoryCollection | null; }
      set categories(value: CategoryCollection | null) { this.set('categories', value); }

      static list(): ArticleCollection {
        return super.list() as unknown as ArticleCollection;
      }

      static async get(id: string, ...includes: ("author" | "categories")[]): Promise<Article>;

      static async get(id: string, ...includes: string[]): Promise<Article> {
        return super.get(id as any, ...includes) as unknown as Promise<Article>;
      }

      static async find(query: Partial<ArticleQuery>, ...includes: ("author" | "categories")[]): Promise<Article>;

      static async find(query?: Partial<ArticleQuery>): Promise<Article> {
        return super.find(query as Record<string, unknown>) as unknown as Promise<Article>;
      }

      static async create(props?: ArticleCreate): Promise<Article> {
        return super.create(props as unknown as Record<string, unknown>) as unknown as Promise<Article>;
      }

      async save(props?: ArticleEdit): Promise<void> {
        await super.save(props as Record<string, unknown>);
      }

      async edit(relationship: 'author', resource: User | number): Promise<void>;
      async edit(relationship: string, resource: unknown): Promise<void> {
        await super.edit(relationship, resource);
      }

      async add(relationship: 'categories', ...resources: (Category | number)[]): Promise<void>;
      async add(relationship: string, ...resources: unknown[]): Promise<void> {
        await super.add(relationship, ...resources);
      }

      async remove(relationship: 'categories', ...resources: (Category | number)[]): Promise<void>;
      async remove(relationship: string, ...resources: unknown[]): Promise<void> {
        await super.remove(relationship, ...resources);
      }

      async reset(relationship: 'categories', ...resources: (Category | number)[]): Promise<void>;
      async reset(relationship: string, ...resources: unknown[]): Promise<void> {
        await super.reset(relationship, ...resources);
      }
}

export class UserCollection extends Collection<User> {
  filter(kwargs: Partial<UserQuery>): this {
    return super.filter(kwargs as Record<string, unknown>);
  }
  fields(params: { users?: ('id' | 'username' | 'articles')[] | null }): this {
    const converted: Record<string, string[]> = {};
    for (const [k, v] of Object.entries(params)) {
      if (v != null) converted[k] = v;
    }
    return super.fields(converted);
  }
}
export class User extends Resource {
      static _type = "users";
      static _attributeTypes: Record<string, string> = {
        'id': 'number',
        'username': 'string',
      };
      static _relationshipTypes: Record<string, [string, boolean]> = {
        'articles': ['articles', true],
      };
      static _capabilities = new Set(['get_many', 'get_one'] as const);
      get username(): string { return this.get('username') as string; }
      set username(value: string) { this.set('username', value); }
      get articles(): ArticleCollection | null { return this.get('articles') as ArticleCollection | null; }
      set articles(value: ArticleCollection | null) { this.set('articles', value); }

      static list(): UserCollection {
        return super.list() as unknown as UserCollection;
      }

      static async get(id: string, ...includes: "articles"[]): Promise<User>;

      static async get(id: string, ...includes: string[]): Promise<User> {
        return super.get(id as any, ...includes) as unknown as Promise<User>;
      }

      static async find(query: Partial<UserQuery>, ...includes: "articles"[]): Promise<User>;

      static async find(query?: Partial<UserQuery>): Promise<User> {
        return super.find(query as Record<string, unknown>) as unknown as Promise<User>;
      }
}

export class CategoryCollection extends Collection<Category> {
  filter(kwargs: Partial<CategoryQuery>): this {
    return super.filter(kwargs as Record<string, unknown>);
  }
  fields(params: { categories?: ('id' | 'name' | 'created_at' | 'articles')[] | null }): this {
    const converted: Record<string, string[]> = {};
    for (const [k, v] of Object.entries(params)) {
      if (v != null) converted[k] = v;
    }
    return super.fields(converted);
  }
}
export class Category extends Resource {
      static _type = "categories";
      static _attributeTypes: Record<string, string> = {
        'id': 'number',
        'name': 'string',
        'created_at': 'string | null',
      };
      static _relationshipTypes: Record<string, [string, boolean]> = {
        'articles': ['articles', true],
      };
      static _capabilities = new Set(['create', 'delete', 'get_many'] as const);
      get name(): string { return this.get('name') as string; }
      set name(value: string) { this.set('name', value); }
      get created_at(): string | null { return this.get('created_at') as string | null; }
      set created_at(value: string | null) { this.set('created_at', value); }
      get articles(): ArticleCollection | null { return this.get('articles') as ArticleCollection | null; }
      set articles(value: ArticleCollection | null) { this.set('articles', value); }

      static list(): CategoryCollection {
        return super.list() as unknown as CategoryCollection;
      }

      static async find(query: Partial<CategoryQuery>, ...includes: "articles"[]): Promise<Category>;

      static async find(query?: Partial<CategoryQuery>): Promise<Category> {
        return super.find(query as Record<string, unknown>) as unknown as Promise<Category>;
      }

      static async create(props?: CategoryCreate): Promise<Category> {
        return super.create(props as unknown as Record<string, unknown>) as unknown as Promise<Category>;
      }
}

File: articles_sdk/resources.py
from __future__ import annotations

import datetime
from typing import Any, ClassVar, Literal, Self, TypedDict, Unpack, cast, overload

from ._runtime.collection import Collection
from ._runtime.resource import Resource


class ArticleQuery(TypedDict, total=False):
    title__contains: str
    category: int | None


class ArticleGetQuery(TypedDict, total=False):
    pass


class ArticleEdit(TypedDict, total=False):
    title: str | None
    content: str | None
    author: User | int | None


class UserQuery(TypedDict, total=False):
    username: str


class UserGetQuery(TypedDict, total=False):
    pass


class CategoryQuery(TypedDict, total=False):
    article: int | None


class ArticleCollection(Collection["Article"]):
    def filter(self, **kwargs: Unpack[ArticleQuery]) -> Self:
        return super().filter(**kwargs)
    def sort(self, *fields: Literal['title', '-title', 'created_at', '-created_at']) -> Self:
        return super().sort(*fields)
    def fields(self, *, articles: list[Literal['id', 'title', 'content', 'created_at', 'author', 'categories']] | None = None) -> Self:
        kwargs = {}
        if articles is not None:
            kwargs['articles'] = articles
        return super().fields(**kwargs)


class UserCollection(Collection["User"]):
    def filter(self, **kwargs: Unpack[UserQuery]) -> Self:
        return super().filter(**kwargs)
    def fields(self, *, users: list[Literal['id', 'username', 'articles']] | None = None) -> Self:
        kwargs = {}
        if users is not None:
            kwargs['users'] = users
        return super().fields(**kwargs)


class CategoryCollection(Collection["Category"]):
    def filter(self, **kwargs: Unpack[CategoryQuery]) -> Self:
        return super().filter(**kwargs)
    def fields(self, *, categories: list[Literal['id', 'name', 'created_at', 'articles']] | None = None) -> Self:
        kwargs = {}
        if categories is not None:
            kwargs['categories'] = categories
        return super().fields(**kwargs)


class Article(Resource):
    _type: ClassVar[str] = "articles"
    _attribute_types: ClassVar[dict[str, Any]] = {
        "id": int,
        "title": str,
        "content": str,
        "created_at": datetime.datetime,
    }
    _relationship_types: ClassVar[dict[str, tuple[str, bool]]] = {
        "author": ("users", False),
        "categories": ("categories", True),
    }
    _capabilities: ClassVar[frozenset[str]] = frozenset(
        {"create", "delete", "edit", "get_many", "get_one"}
    )
    _relationship_capabilities: ClassVar[dict[str, frozenset[str]]] = {
        "author": frozenset({"edit"}),
        "categories": frozenset({"add", "remove", "reset"}),
    }
    _collection_class: ClassVar = ArticleCollection

    id: int
    title: str
    content: str
    created_at: datetime.datetime
    author: User
    categories: CategoryCollection

    @classmethod
    def list(cls) -> ArticleCollection:
        return cast(ArticleCollection, super().list())

    @overload
    @classmethod
    async def get(cls, id: int | str, *includes: Literal["author", "categories"]) -> Article: ...

    @overload
    @classmethod
    async def get(cls, id: int | str, *includes: str) -> Article: ...
    @classmethod
    async def get(cls, id: int | str | None = None, *includes: str, **query: Any) -> Article:
        return cast(Article, await super().get(id, *includes, **query))

    @overload
    @classmethod
    async def find(cls, *includes: Literal["author", "categories"], **query: Unpack[ArticleQuery]) -> Article: ...

    @overload
    @classmethod
    async def find(cls, *includes: str, **query: Unpack[ArticleQuery]) -> Article: ...
    @classmethod
    async def find(cls, *includes: str, **query: Unpack[ArticleQuery]) -> Article:
        return cast(Article, await super().find(*includes, **query))

    @classmethod
    async def create(cls, *, title: str, content: str, author: User | int | None = None, categories: list[Category | list[int] | None] | None = None) -> Article:
        kwargs: dict[str, Any] = {"title": title, "content": content}
        if author is not None:
            kwargs["author"] = author
        if categories is not None:
            kwargs["categories"] = categories
        return cast(Article, await super().create(**kwargs))

    async def save(
        self, *fields: str, force_create: bool = False, **kwargs: Unpack[ArticleEdit]
    ) -> None:
        await super().save(*fields, force_create=force_create, **kwargs)

    @overload
    async def edit(self, relationship: Literal["author"], resource: User | int) -> None: ...
    @overload
    async def edit(self, relationship: str, resource: Any) -> None: ...
    async def edit(self, relationship: str, resource: Any) -> None:
        return await super().edit(relationship, resource)

    @overload
    async def add(self, relationship: Literal["categories"], *resources: Category | int) -> None: ...
    @overload
    async def add(self, relationship: str, *resources: Any) -> None: ...
    async def add(self, relationship: str, *resources: Any) -> None:
        return await super().add(relationship, *resources)

    @overload
    async def remove(self, relationship: Literal["categories"], *resources: Category | int) -> None: ...
    @overload
    async def remove(self, relationship: str, *resources: Any) -> None: ...
    async def remove(self, relationship: str, *resources: Any) -> None:
        return await super().remove(relationship, *resources)

    @overload
    async def reset(self, relationship: Literal["categories"], *resources: Category | int) -> None: ...
    @overload
    async def reset(self, relationship: str, *resources: Any) -> None: ...
    async def reset(self, relationship: str, *resources: Any) -> None:
        return await super().reset(relationship, *resources)


class User(Resource):
    _type: ClassVar[str] = "users"
    _attribute_types: ClassVar[dict[str, Any]] = {
        "id": int,
        "username": str,
    }
    _relationship_types: ClassVar[dict[str, tuple[str, bool]]] = {
        "articles": ("articles", True),
    }
    _capabilities: ClassVar[frozenset[str]] = frozenset(
        {"get_many", "get_one"}
    )
    _collection_class: ClassVar = UserCollection

    id: int
    username: str
    articles: ArticleCollection

    @classmethod
    def list(cls) -> UserCollection:
        return cast(UserCollection, super().list())

    @overload
    @classmethod
    async def get(cls, id: int | str, *includes: Literal["articles"]) -> User: ...

    @overload
    @classmethod
    async def get(cls, id: int | str, *includes: str) -> User: ...
    @classmethod
    async def get(cls, id: int | str | None = None, *includes: str, **query: Any) -> User:
        return cast(User, await super().get(id, *includes, **query))

    @overload
    @classmethod
    async def find(cls, *includes: Literal["articles"], **query: Unpack[UserQuery]) -> User: ...

    @overload
    @classmethod
    async def find(cls, *includes: str, **query: Unpack[UserQuery]) -> User: ...
    @classmethod
    async def find(cls, *includes: str, **query: Unpack[UserQuery]) -> User:
        return cast(User, await super().find(*includes, **query))


class Category(Resource):
    _type: ClassVar[str] = "categories"
    _attribute_types: ClassVar[dict[str, Any]] = {
        "id": int,
        "name": str,
        "created_at": datetime.datetime | None,
    }
    _relationship_types: ClassVar[dict[str, tuple[str, bool]]] = {
        "articles": ("articles", True),
    }
    _capabilities: ClassVar[frozenset[str]] = frozenset(
        {"create", "delete", "get_many"}
    )
    _collection_class: ClassVar = CategoryCollection

    id: int
    name: str
    created_at: datetime.datetime | None
    articles: ArticleCollection

    @classmethod
    def list(cls) -> CategoryCollection:
        return cast(CategoryCollection, super().list())

    @overload
    @classmethod
    async def find(cls, *includes: Literal["articles"], **query: Unpack[CategoryQuery]) -> Category: ...

    @overload
    @classmethod
    async def find(cls, *includes: str, **query: Unpack[CategoryQuery]) -> Category: ...
    @classmethod
    async def find(cls, *includes: str, **query: Unpack[CategoryQuery]) -> Category:
        return cast(Category, await super().find(*includes, **query))

    @classmethod
    async def create(cls, *, name: str) -> Category:
        kwargs: dict[str, Any] = {"name": name}
        return cast(Category, await super().create(**kwargs))
